R/weighted_efficacy.R

```diff
@@ -0,0 +1,275 @@
+  lapply(dose_times, function(x){
+    had_first_beyond_threshold <- t - x[, 1] >= dose_period
+    had_first_beyond_threshold[is.na(had_first_beyond_threshold)] <- FALSE
+    not_had_2nd <- is.na(x[, 2])
+    had_first_beyond_threshold & not_had_2nd
+  })
+}
+
+#' Administer first doses
+#'
+#' @param ad Number of doses to administer in each age group
+#' @param dose_times Dose monitoring data
+#' @param t Current time
+administer_first_dose <- function(ad, dose_times, t){
+  for(i in seq_along(ad)){
+    if(ad[i] > 0){
+      to_jab <- which(is.na(dose_times[[i]][,1]))[1:ad[i]]
+      dose_times[[i]][to_jab,1] <- t
+    }
+  }
+  return(dose_times)
+}
+
+#' Administer second doses
+#'
+#' @param ad Number of doses to administer in each age group
+#' @param dose_times Dose monitoring data
+#' @param dose_period Days between dose 1 and dose 2
+#' @param t Current time
+administer_second_dose <- function(ad, dose_times, dose_period, t){
+  eligable <- eligable_for_second(dose_times, t, dose_period)
+  for(i in seq_along(ad)){
+    if(ad[i] > 0){
+      to_jab <- which(eligable[[i]])[1:ad[i]]
+      dose_times[[i]][to_jab,2] <- t
+    }
+  }
+  return(dose_times)
+}
+
+
+#' Extract dose numbers administers
+#'
+#' @param dose_times Dose monitoring data
+#' @inheritParams weighted_efficacy
+extract_dose_number <- function(dose_times, maxt){
+  output_1dose <- matrix(0, nrow = maxt, ncol = length(dose_times))
+  colnames(output_1dose) <- 1:length(dose_times)
+  output_2dose <- matrix(0, nrow = maxt, ncol = length(dose_times))
+  colnames(output_2dose) <- 1:length(dose_times)
+  for(t in 1:maxt){
+    for(a in 1:length(dose_times)){
+      output_1dose[t, a] <- sum(dose_times[[a]][,1] == t, na.rm = TRUE)
+      output_2dose[t, a] <- sum(dose_times[[a]][,2] == t, na.rm = TRUE)
+    }
+  }
+  output_1dose <- as.data.frame(output_1dose) %>%
+    dplyr::mutate(t = 1:maxt) %>%
+    tidyr::pivot_longer(-.data$t, names_to = "age_group", values_to = "n_1dose") %>%
+    dplyr::mutate(age_group = factor(.data$age_group, levels = 1:length(dose_times)))
+  output_2dose <- as.data.frame(output_2dose) %>%
+    dplyr::mutate(t = 1:maxt) %>%
+    tidyr::pivot_longer(-.data$t, names_to = "age_group", values_to = "n_2dose") %>%
+    dplyr::mutate(age_group = factor(.data$age_group, levels = 1:length(dose_times)))
+  output <- dplyr::left_join(output_1dose, output_2dose, by = c("t", "age_group"))
+  return(output)
+}
+
+
+#' Extract weighted efficacies
+#'
+#' @param dose_number  Dose number: 1 or 2
+#' @inheritParams weighted_efficacy
+add_weighted_efficacy <- function(dose_number, infection_efficacy, disease_efficacy, v1v2){
+  dose_number  %>%
+    dplyr::group_by(.data$age_group) %>%
+    dplyr::mutate(c2 = cumsum(.data$n_2dose),
+                  c1 = cumsum(.data$n_1dose) - .data$c2) %>%
+    # Number vaccine protected - lag between administration of dose 1 and protection
+    dplyr::mutate(protected1 = dplyr::lag(.data$c1, v1v2, default = 0),
```

Comment:
This line I think is what is causing the intermediate values for efficacy I talked about as the lag means you can end up having all of your population in protected1 as well as some people in protected 2 at the same time if they manage to give all second doses quicker than v1v2. E.g. with the current code: 

![image](https://user-images.githubusercontent.com/15249565/120389614-8dca0c80-c324-11eb-864e-443cd89d86a0.png)

But with the following you woudl now take the lagged value but subtract those that are in protected2 already and catch for negatives. The following has the slight issue of saying people have started receiving second doses before `v1v2` has occurred.

```suggestion
    dplyr::mutate(protected2 = .data$c2,
                          protected1 = dplyr::lag(.data$c1, v1v2, default = 0) - .data$c2) %>%
    dplyr::mutate(protected1 = replace(.data$protected1, .data$protected1 < 0, 0)) %>%
```

But it does give the correct increase in efficacy over time:

![image](https://user-images.githubusercontent.com/15249565/120391085-755af180-c326-11eb-8f0c-a259c75bbf55.png)

(let me know if you need code to reproduce these)